### ISSUE TITLE
Datagram vs stream & relative stream prioritization

### ISSUE THREAD

**Issue by vasilvv** [w3c/webtransport#62]
Currently, we specify no particular order in which stream data and datagrams should go.

The way the current scheme works in Google's implementation is roughly:
1. A timer or a received packet causes the connection to become writable.
2. The connection notifies the application so that it can send its datagrams.
3. The connection checks the send buffers for the open streams and sends stream data.

This works only if all operations are synchronous, and we can't make step 2 synchronous since the JS callback would be asynchronous.  Thus, we need to come up with a way to provide an opening for the JS to send datagrams while there's always stream data available to send.

A "version 0" approach would be to add fixed size buffer of datagram that gets always drained before stream data.  That works, but if the buffer is too small, it can hurt performance, and if it's too big, this can lead to overbuffering.

One possible improvement would be to dynamically size the datagram buffer based on the congestion control window or some other criteria.  One could possibly let the user set that buffer size manually.  There is still, however, an implicit race between the JS application adding datagrams into the buffer and the network process deciding to send pending stream data instead.

Another idea I had is that the user can "reserve" a portion of congestion window for datagrams and leave the rest for stream data.  That would look roughly like this: a user asks for a "token" that reserves a portion of the congestion window aside for a datagram.  When the user sends the datagram, they return that token, since the reserved portion of the window is now used by the newly sent datagram.

**Comment by Atrius** [comment]
I can reserve datagram tokens even if the cwnd is full, right?

Otherwise, I think there's the same race, just on the feedback side:  whenever an ack arrives, the transport has to fire a callback to the app telling it that there's more room in the cwnd, and in the meantime, it can use that space to send stream data.

I could imagine this working pretty well for my use case, which is real-time media.  I imagine I would just keep about 1 frame's worth of datagram tokens on hand, so that I can send the next frame as soon as it's ready, and any excess cwnd space can be used for stream data.

**Comment by vasilvv** [comment]
Yup, that's the idea.  I imagine that some applications would want to adjust the reservation based on the current CWND size, but without knowing specifically how those look like, I don't know what the API should be.

**Comment by jan-ivar** [comment]
Meeting:
- Fyi:, Chrome always attempts to send datagrams before stream data.

**Comment by martinthomson** [comment]
My sense here is that we want to have a priority API that would allow us to manage relative priorities.

There might be a case here for having multiple datagram queues in order to allow datagrams to operate at different priorities. An alternative might be to allow the application to manage changing datagram priority.

**Comment by yutakahirano** [comment]
@vasilvv can you state yout current thoughts, as this issue is very old?

I don't think you're planning to introduce the feature for the initial release (please correct me if I'm wrong), but it's possible that we need to change our current specified APIs to support this use-case.

One thing I want to clarify is whether we want to handle datagrams of various sizes.

If we need to handle only one size, assuming there were a good prirority API, we can cover the use case with [`DatagramDuplexStream.outgoingHighWaterMark`](https://w3c.github.io/webtransport/#dom-datagramduplexstream-outgoinghighwatermark). 

If we need to handle datagrams of various sizes, the current `DatagramDuplexStream.outgoingHighWaterMark`](https://w3c.github.io/webtransport/#dom-datagramduplexstream-outgoinghighwatermark) is insufficient because it counts the number of datagrams. We need something that counts the number of bytes of queued datagrams. This can affect the discussion at #305 as we may want to make the APIs more flexible, before the initial release.

@ricea FYI

**Comment by jan-ivar** [comment]
Meeting:
- Maybe prioritize datagrams over streams? But if you have a pull-based source that might swamp non-datagrams.
- Struggling to define language around how user agents should work here.

**Comment by jan-ivar** [comment]
Meeting:
- Once #340 is merged, we plan to change the Milestone of this to sometime in the future. Basically this says datagrams take priority over other (webtransport stream) traffic, until further notice. We may come back to add more detailed prioritization APIs later.

**Comment by jan-ivar** [comment]
Meeting:
- @vasilvv & @ddragana mentioned related work https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-priority but is on the http level not stream level. Uses headers with a number 0-7 where 0 is highest. 
  - Split off due to it going slower due to questions over whether it is useful in practice
  - In http it's just a suggestion to the server how to prioritize, which is different from WebTransport which also would need to affect client sending.
 - Revisit once we have more developer feedback.

**Comment by LPardue** [comment]
FWIW, the Extensible priorities spec allows the priority information to be used to influence local client decisions, but that's a client implementation choice and there's really not much more to say at the protocol level.

The I-D https://datatracker.ietf.org/doc/html/draft-pardue-masque-dgram-priority is a response to people asking how Extensible Priorities might extend to Datagrams. This is not adopted by any group (yet?) and illustrates how this scheme can be extended to support a different data-bearing type alongside streams.

The high-order question though really is, is there any Priority API that allows expressing send and receive priority of fully duplex flows. Without that, you bake in some assumptions.

**Comment by jan-ivar** [comment]
So if you're the producer of data being written to the WT sender, you can throttle that way per stream.

But I'm not sure we made the right decision that datagrams should always trump streams. One example would be a case where media is sent over datagrams but with a stream control channel. We don't want to swamp the control channel.

May I prose we at least add a boolean to flip this? Maybe
```js
wt.datagrams.priorityOverStreams = true (default) | false
```
...except the web platform apparently frowns on booleans that default to true, so how about:
```js
wt.datagrams.priority = "high" | "low"
```
?

**Comment by martinthomson** [comment]
Whatever we do here should be made consistent with whatever we do for stream priorities.  Why not do the same for both?  There might be a natural advantage given to datagrams over streams, or a different default, but any deliberate choice should deal with the potential for starvation (inter-stream, inter-session, stream-vs-datagram, etc...).

**Comment by LPardue** [comment]
In my (non-WebTransport) experience, there always a fairly decent chance of picking the wrong data type to be sent first and causing starvation. This becomes more apparent in more complex or bandwidth-intensive workloads . Some kind of timeshare is required to avoid that. In draft-pardue-masque-dgram-priority, the simple suggestion I made was a 50/50 share between streams and datagrams that were deemed equal priority.

**Comment by jan-ivar** [comment]
FYI, in the current W3C spec (and Chrome implementation) sent datagrams have [100% priority](https://w3c.github.io/webtransport/#sendstream-write) over streams: _"7. Wait for transport’s [[[Datagrams]]](https://w3c.github.io/webtransport/#webtransport-datagrams)'s [[[OutgoingDatagramsQueue]]](https://w3c.github.io/webtransport/#datagramduplexstream-outgoingdatagramsqueue) to be empty. Note: This means outgoing datagrams are prioritized over stream data."_

**Comment by martinthomson** [comment]
I think that 100% is inadvisable, but the general idea of datagrams pre-empting streams is fine.

Overall though, I'm concerned that a point solution for the starvation problem, without considering inter-stream prioritization as part of that, might not be the best way to approach this.  Prioritization of streams into a small number of slots, with datagram priorities interleaved between those is probably where I would look here.  I'm not sure about how easy it would be to reliably implement a 50/50 share in our implementation.

**Comment by LPardue** [comment]
Yeah no need for hard proportions here. What has worked for me is, when preparing a packet to send, look at queues of stream data and datagram messages (slots if you like), and alternate between which one to pop off. If the preferred type is empty, move onto the other for this round to avoid wasted capacity.

**Comment by aboba** [comment]
Mathis Engelbart has looked at impact of prioritizing datagrams within a QUIC connection operating with NewReno.  This results in high datagram latency, since the reliable stream fills the congestion window when datagrams aren't being sent. 
Results are [here](https://docs.google.com/presentation/d/16gxsLckg7mLzYhIiQm0wPgUJHXNPfivbTmln-hhxlJw/edit#slide=id.g1143eb38147_6_224).

**Comment by jan-ivar** [comment]
I've filed https://github.com/w3c/webtransport/issues/388 on a flaw that might explain some of that.

**Comment by wilaw** [comment]
See related question on receiver throttling at #389

**Comment by jan-ivar** [comment]
Bikeshedding:
```js
stream.priority = 1; // is this highest or lowest priority? Or do we start at 0? And how many levels?
```
or do we come up with enough enums?
```js
stream.priority = "lowest" | "very low" | "low" | "medium" | "high" | "very high" | "highest"
```

**Comment by wilaw** [comment]
@vasilvv suggested during last call that a pair of Int 32 attributes  might be preferable for setting stream priority. Can you provide further information on this proposal? 

Also a question : assume a numerical priority scheme where 0 is default and any bigger number implies higher priority. Two streams start writing - stream A with a priority of 0 and stream B with a priority of 1. Does all of stream B need to be sent before any of stream A is sent, or is their sending multiplexed with some bias towards having more of stream B sent than A? If so, how is the relative send ratio determined by the user-agent?

**Comment by jan-ivar** [comment]
The only use case I've heard for that many priority levels is a frame-per-stream solution with a "newest frame first" strategy: give each new frame a higher priority level than the last, to create a LIFO queue of sorts.

But potential problems with such an approach aside (backlog of old frames to be sent, or nothing being sent at all), this use case seems solved without priorities today by `abort`ing sending of earlier frames whenever new ones become available:
```js
async function sendFrame(wt, encodedFrame) {
  const perFrameStream = await wt.createUnidirectionalStream();
  const writer = perFrameStream.getWriter();
  writer.write(encodedFrame).catch(() => {});
  this.lastFrameWriter?.abort(); // abort sending last frame if it hasn't been sent or ack'ed yet
  this.lastFrameWriter = writer;
  await writer.close();
}
```
So I don't think a priority scheme should try to solve this.

In my mind, a few levels for JS to play with that block lower-priority streams from sending, seems sufficient. When datagrams are set to a higher priority than other streams, other queues should pause whenever new datagrams need to be sent. This should probably be the default.

**Comment by kixelated** [comment]
Hey folks, I published a [draft](https://www.ietf.org/id/draft-lcurley-warp-00.html) outlining how Twitch is using QUIC for live media transport. Prioritization is crucially important and is _the_ reason why we're using QUIC and WebTransport streams. 

Warp works by prioritizing audio and newer streams first. The idea is that older video streams will starve so the player will skip over them instead of buffering. 

This is quite difficult to do with a handful of levels (ex. 0-7) because new streams are created constantly while old streams could starve for a significant period of time. It's possible given a way to reset stream priority, but this is complicated and racy. 

We also want to specify that higher priority streams will be sent first (except retransmissions?). It's not clear how a "high" stream interacts with a "medium" stream. For Warp, we want any available data for stream priority 642 to be sent before stream priority 641. I call this priority scheme `precedence` and it's the simplest to implement, although it can cause starvation if not done properly.

We're currently using Warp to push media from the server to the WebTransport client. This is implemented on the server by assigning an integer priority to each stream. The QUIC implementation will loop through streams in descending priority order when trying to populate a packet with STREAM frames.

We'll need a way of specifying stream prioritization via WebTransport in order to support Warp browser ingest. Victor's recommendation of a single uint32 would work great for Warp assuming streams are prioritized like I described. I would recommend the ability to specify the prioritization scheme (at the connection level?) to remove this ambiguity.

**Comment by kixelated** [comment]
@jan-ivar Abandoning streams works in some situations but it's not prioritization. For my live media use-case, we want to deliver the newest data first but we don't want to drop older media because it might still arrive in time.

For example, one viewer might have a 2s buffer. There might be a momentary network blip that causes a gap in the buffer (because we sent the newest stream/frames first) but the network might recover in time to backfill the buffer. If we abandon streams at the first sign of congestion, the media is lost forever.

**Comment by jan-ivar** [comment]
Meeting:
- Requests appear to be for strict prioritization
- Victor: 2 int32s: a strict priority + a second weigthed one for ones with same strict priority
- Other Ideas: when creating a stream, option to say prioritize it over all prior ones. @kixelated would that work for your use case?
- Poll QUIC implementations about their interest in doing this?

**Comment by LPardue** [comment]
Regarding 2 int32s. Playing the IETF card here, it seems like it would be better, if possible, to align within the constraints of the extensible priority scheme because that has already run the standards process and is defined for HTTP/2 and HTTP/3.

Ignoring extensible priorities seems to need to invent something similar but slightly different, and the knock on effect is H3 and QUIC stacks would need to implement something similar but different.

So back to 2 uint32s. Do you need that much granularity in each? What is the developer experience here - the Priority hints spec settled for 3 levels. 

Putting the fidelity of numbers aside, the design seems like a scheme that could be supported by extension parameters. For instance, the second uint32 being instead an "order" parameter. This could override the default recommendation to send in order of stream ID.

**Comment by alvestrand** [comment]
In the WebRTC discussions (which wanted to prioritize long-lived media streams together with data transfers of arbitrary shapes), we ended up specifying:

1) Very few levels, as more levels was seen as just confusing
2) A weighted round-robin scheme (WFQ), so that lower priority streams wouldn't get entirely starved

Five years after it was specified, the scheme has still not been implemented; all WebRTC implementations send everything with equal priority as soon as congestion control allows.

The last point may show how important this is...

**Comment by martinthomson** [comment]
In HTTP we have a similar experience.  Even a modestly complex scheme was not implemented consistently.  We're still not completely certain about the simpler one.  Though we have some implementation, what we really want to confirm that it is good is widespread and consistent implementation so that people can come to rely on it.

**Comment by LPardue** [comment]
I agree that extensible priorities remains to be proven, good observation.

Given that along with what Harald said, is it too early to charge headlong into any choice here?

**Comment by jan-ivar** [comment]
> This could override the default recommendation to send in order of stream ID.

Is stream ID determined by stream creation order? If so, what happens if JS creates 1000 streams, and starts sending them in reverse order? Might this work as a workaround here?

**Comment by LPardue** [comment]
RFC 9000 says in https://datatracker.ietf.org/doc/html/rfc9000#section-2.1

> A stream ID that is
   used out of order results in all streams of that type with lower-
   numbered stream IDs also being opened.
   
and in https://datatracker.ietf.org/doc/html/rfc9000#section-3.2

>    Before a stream is created, all streams of the same type with lower-
   numbered stream IDs MUST be created.  This ensures that the creation
   order for streams is consistent on both endpoints.

**Comment by kixelated** [comment]
The ability to specify both a strict and weighted priority is an interesting idea. However, I can't envision a situation where it would be necessary to specify both.

During congestion, I want some ensure that important resources are fully transferred first. It's typically not useful to partially transfer all resources, or partially transfer some resources more than others. This is especially true when dependencies are involved.

For that reason, I don't think relative weights are useful. That lines up with some examples people have mentioned, where it hasn't even been necessary to implement. I understand the rationale: starvation is scary, but it's just necessary to reduce concurrency in response to congestion.

**Comment by kixelated** [comment]
As for stream order determining transfer order, it would feel like a hack. Like @jan-ivar mentioned, you would need to create a pool of streams and selectively use them, hoping that you don't run into flow control limits in the process. I think it would be much easier to provide an integer from `0-n` rather than constantly drain and refill a pool of `n` streams.

I put two examples for how I would set the uint32 in the [warp draft](https://www.ietf.org/archive/id/draft-lcurley-warp-00.html#name-live-content). In production we're using something slightly more advanced so it works with discontinuities. For example, watching a live stream preferring newer content first, then switching to an advertisement preferring the data in order (dropping video is disabled).

**Comment by kixelated** [comment]
I would recommend `stream.setOrder(int32)`. It's easy to implement and you could always add another form of prioritization in the future via a new method.

Alternatively, `stream.dependsOn(otherStream)` is a more generic way to perform "strict prioritization" without a single number space. That method name is confusing though because both streams still behave independently.

**Comment by LPardue** [comment]
An equivalent for order proposal could be to use extensible priority parameters dict. So something like

let priority = {order: int32}:
stream.setPrioriry(priority);

This is equivalent to default urgency and default incremental (I.e. everything in the same bucket) so that everything can be ordered.

I would recommend staying away explicit dependencies wherever possible.

**Comment by jan-ivar** [comment]
Taking a step back, I'm hearing these use cases for setting strict priorities:
 1. Large file uploads in the background should never degrade live media ingest
 2. Never starve my control channel or datagrams
 3. Ad hoc congestion control

The first two seem solved by three levels: control channel = `"high"`, media = `"medium"`, file upload = `"low"` 

But to solve the third we're up to INT32_MAX, which is why I wonder: should we solve congestion in https://github.com/w3c/webtransport/issues/365 or here?

If here: is priority the right API to express congestion handling? Whereas WebRTC's solution to maintain timeliness is to reduce video resolution or frame-rate, Warp's appears to be to skip occasional video (not audio) segments.

But using increasing stream-priorities to express this intent seems sub-optimal in that the user agent doesn't know what's going on, which could cause starvation or accumulate streams (and related JS objects) that never get sent or GC'ed.

I also wonder how this fits with the QUIC spec's assumptions around use (and resources): e.g. QUIC seems to assume streams are created when data is available to be sent, and that [_"stream flow control credit has successfully been reserved to send the written data"_](https://datatracker.ietf.org/doc/html/rfc9000#section-2.4). Not sure how that works, but the QUIC spec at least doesn't seem to deter a naive implementation from deciding each new stream ID might represent some interest in bandwidth. If the app aborts them after a while, and also avoids starvation, then this could no doubt work, but is it the most robust API?

> If we abandon streams at the first sign of congestion, the media is lost forever and it causes a poor user experience.

If this is [to support rewind](https://www.ietf.org/id/draft-lcurley-warp-00.html#name-prioritization) could the missed segments be re-requested in a low-priority stream instead?

**Comment by jan-ivar** [comment]
How about a simple `priority` boolean? I.e. when you create a new stream you can say "this stream is a priority", which makes it take precedence over all previously created streams. If you don't (the default) then the stream is prioritized after all previously created streams. @kixelated would this suffice your use case?

**Comment by jan-ivar** [comment]
And I forgot datagrams always get priority, because they don't make much sense otherwise.

**Comment by kixelated** [comment]
Warp generally prioritizes newer streams, but using LIFO would be a hack. We would need to recreate streams rather aggressively just to pump them to the top of the stack. For example, sending a PAUSE message over an API channel must always be the highest priority, as it would be a horrible user experience if it was starved behind megabytes of video data.

LIFO would definitely not work for other video delivery protocols. For example, [RUSH](https://www.ietf.org/archive/id/draft-kpugin-rush-00.html) delivers each frame as a unique stream. Video is encoded as deltas; generally each frame depends on the previous frame, and we need to match that dependency at the network layer. That sounds like FIFO, however there are periodic I-frames that do not depend on previous frames (effectively a PNG). Media protocols like Warp, RUSH, WebRTC, etc can drop media by skipping ahead to an I-frame rather than try to delivery everything in order (HLS, RTMP, etc).

**Comment by kixelated** [comment]
Here's a quick summary of how (live) video works.

## Encoding
I-frames do not depend on other frames while P-frames depend on one or more previous frames. The decoder will crash or produce artifacts if these dependencies are not maintained.

![1](https://user-images.githubusercontent.com/432854/169910737-620ec7f5-9206-4430-9051-afca4858aeeb.png)

## HLS/DASH
HLS and DASH split video into segments that start with an I-frame. These segments are downloaded sequentially and are virtually never dropped.

This is effectively FIFO, as the oldest segments are prioritized. Higher numbers are delivered first, in strict order.

![5](https://user-images.githubusercontent.com/432854/169910967-4c4df8e1-fe79-4167-9ac9-5a8b9c1d8dfe.png)

## Warp
Warp is simple; we flip the priority around such that the newest segment is delivered first.

This is basically LIFO, but it gets more complicated when audio and metadata is involved. Warp uses the fact that streams are delivered in order to maintain P-frame ordering. Same as before, higher numbers are delivered first, in strict order.

![2](https://user-images.githubusercontent.com/432854/169911446-3b42fbef-c38e-4de5-a4b7-4225b2bd0d59.png)

## RUSH/WebRTC
RUSH and WebRTC deliver each frame independently, but still need to enforce some form of frame ordering during congestion. 

This is normally done by aggressively dropping frames rather than retransmit, but it can be generalized using prioritization.

![3](https://user-images.githubusercontent.com/432854/169911606-16930707-e5ea-410a-9136-56ccf29d9f48.png)

## Encoding v2
Higher quality media will use b-frames, which annoyingly reference one or more future frames and zero or more past frames.  In an ideal world, these media dependencies could be mapped to network dependencies. Here's an example but there are many different types of GOP structures:

![4](https://user-images.githubusercontent.com/432854/169912068-4867ab81-946b-4905-a047-61d03780dc3f.png)

WebRTC doesn't use these because they introduce latency. Warp currently delivers segments in decode order, and I don't know what RUSH does. It's an open problem as how to best map these types of frames to QUIC.

**Comment by LPardue** [comment]
Thanks Luke.

To explain extensible priorities in similar terms, the spec recommends something that, when all streams are non-incfemental, equates to 8 FIFO queues serving stream IDs lower to higher. It would be fairly trivial to add some extension parameter to make the ordering of these queues more configurable (either a blanket change  per-queue change, or explicit order based on some number other than stream ID).

**Comment by jan-ivar** [comment]
Meeting:
- int32 priority is definitely one of the options to consider.
- Some concerns still that starvation and buffer bloat are paper cuts that might afflict apps owning this problem
- Implementation complexity is unclear.
- Would it be fair for an implementer to explore this in a trial, so we can learn if this works well?

**Comment by martinthomson** [comment]
So it looks like @LPardue is pointing at the decision made in HTTP to establish [a common baseline of 8 FIFO queues](https://httpwg.org/http-extensions/draft-ietf-httpbis-priority.html).  Our experience there was that more complex priority signaling schemes were not consistently implemented, or not implemented at all in some cases.

At some level, any number of FIFOs can be used to implement the warp model (HLS works with a single priority that operate on a FIFO basis).  Depending on how many streams are outstanding, you need to reallocate priorities as your supply of priorities exhaust.  At just 8, it might be that the rate at which you need to perform this reshuffle is pretty high.  This is worse if you consider the need to maintain separate priorities for audio and video within that same space.  So I can appreciate how many $2^3$ might be a little too parsimonious.

I'm extremely skeptical of the prospects for more complex signaling, which includes schemes that includes more numbers.  @vasilvv makes a compelling argument for the browser taking the complexity hit.  Unlike in HTTP, we have the advantage of a relatively small set of implementers, with a relatively high tolerance for complexity (I say that like it is a good thing, which feels dirty).  It's possible that a more complex interface is appropriate here.  But I'd be uncomfortable committing to something more complex without first trying it on.

**Comment by LPardue** [comment]
I think that's a close summary of what I'm saying.

> (HLS works with a single priority that operate on a FIFO basis)

FWIW I've seen first hand problems caused by manifest updates being blocked behind segment requests. So even there different queues can help.

> Depending on how many streams are outstanding, you need to reallocate priorities as your supply of priorities exhaust. At just 8, it might be that the rate at which you need to perform this reshuffle is pretty high.

I think that holds if you ignore the order dimension. However, within a single extensible priority queue, the ordering is _recommended_ to be based on stream ID. So my thinking is you can have a 8 coarse buckets of $thing (say based on media type or control data) and within that you can deterministically dispatch individual items based on an ordering dimension. For use case where FIFO is not preferred, some other thing (like a new `order` param) can modulate dispatch behaviour.

Using a single value like uint32 for priority would immediately cause me to ask - "What is the expected behaviour when two items have the same priority?"

**Comment by alvestrand** [comment]
The HTTP priority scheme seems to default to "resources are set one by one" - it offers the "incremental" flag to communicate that there's an advantage to offering parts of multiple responses.

This is in contrast to the media-transmission situation, where it is normal to have multiple conceptual streams that *must* be interleaved for sanity (frames may still be represented as streams or as datagrams; YMMV)

In the WEBRTC prioriity discussions, the WG ended up specifying a weighted round robin scheme for handling priority - trying to ensure that lower priority streams do not get completely starved out by higher priority schemes - the HTTP priority scheme doesn't seem to be prescriptive in that regard.
(FWIW, as of May 2022, the WRR scheme for webrtc priority has never been implemented.)

**Comment by jan-ivar** [comment]
I'm hearing general agreement for _**at least 3**_ levels of strict priority — by which I only mean no-one seems satisfied by _less_, which means we should continue discussion. Agreement seems to hold up to ~8 levels, and then drops above that. So I think it'd be useful to explore if WARP could be pulled off with fewer levels:

> Depending on how many streams are outstanding, you need to reallocate priorities as your supply of priorities exhaust. At just 8, it might be that the rate at which you need to perform this reshuffle is pretty high.

An overly simplistic example situation relying on int32 would seem to go something like this (assume 1 second segments):
1. I `write()` my first segment A at priority 1 (lowest). I know from the promise when it's been sent
2. But my next segment B is ready sooner, so I `write(B)` with priority 2 before then. This pauses A to send B instead
3. My next segment C after that is _also_ ready sooner, so I `write(C)` with priority 3, pausing B to send C
4. and so on...
5. At some point I cancel streams that are too old to avoid send buffer bloat

Never mind that this describes starvation atm — assume enough time passes eventually to avoid it — my aim is merely to illustrate a situation that relies on increasing priority levels to keep things going.

But look at step 2 where we're _effectively down-prioritizing A_. What if we did this literally instead? Using only 2 levels:
1. I `write()` my first segment A at priority `"medium"`. I know from the promise when it's been sent
2. But segment B is ready sooner, so I set `A.priority = "low"` and `write(B)` at `"medium"`, pausing A to send B
3. Segment C is _also_ ready sooner, so I set `B.priority = "low"` and `write(C)` at `"medium"` pausing B to send C
4. and so on...
5. At some point I cancel streams that are too old to avoid send buffer bloat

This seems comparable in that it always prioritizes the latest 1-second segment (C), and only differs in it's prioritizing of missed media (A vs B) since we only have 2 levels. With 8 levels an app could prioritize up to 7 seconds of missed playback.

The key here is instead of relying on rigidly increasing numbers, we allow the app to change the priority of ongoing streams. @kixelated thoughts?

**Comment by LPardue** [comment]
Changing priorities locally (or simply just prompting existing prioritised data with something more important) can be done quite reactively (aside: in TCP large write buffers can neutralise this, because you're already committed, webtransport should look out for that).

Some of the concerns about reprioritization come about because it's the remote peer that wants reprioritize. When chasing low latency transfer, the latency of reprioritization signals and reaction often work against the goals, with a poor outcome.

**Comment by LPardue** [comment]
Forgot to make the critical point, when yalk about prioritization changes after initiation we need to be clear what exactly to target actor is.

**Comment by jan-ivar** [comment]
Not sure I follow. I was purposely avoiding externalities above, by observing that, all else being equal, rigidly increasing numbers and changeable priorities are two ways to express the same thing: a change in priority of an ongoing stream.

**Comment by LPardue** [comment]
You're not wrong. I'm just highlighting that discussion in other venues has highlighted that changing priorities of open and active streams  whether local or remote, can be too "slow" for some use cases. 

I get that WebTransport API has to focus on its own local needs in the web platform. But it will probably need to be compatible or adaptable with the other ecosystem of clients and servers. So your choice here has effects.

**Comment by martinthomson** [comment]
What Jan-Ivar is describing is effectively what the browser would have to do if the API had more levels.  So it is clearly possible, modulo the part where network issues cause the whole thing to break down.  As Luke says, this is easier on the app if the number of levels >= the number of active streams, but it is possible to have approximations with fewer.

**Comment by jan-ivar** [comment]
> Warp would need at least 4 levels (new audio, new video, prev video, prev audio) to handle RTTs <2s

Great!

> However, RUSH would need at least 120 levels at 30fps given how frequently it creates streams.

Why not 4 levels? Assuming Warp is also 30fps, if there's no new network info for 2 seconds, what's wrong with either Warp or RUSH filling the send buffer with 30 frames at "new video" priority and 30 frames at "prev video" priority however interleaved (and however many audio segments)? Hypothetically, just to set the next point:

Why not 6 levels? To improve on that baseline comparison: If we want to send 30 fps video but anticipate bad congestion (more appropriate for, say, 5fps), why not e.g. send every 3rd frame at +1 priority, and every 6th frame at +2 priority? This would appear to give us 30 fps with fallback to 10fps and then 5fps.

Disclaimer: the above are thought experiments off of presented diagrams, since I know nothing about how RUSH actually works. My point is merely to question whether JS needs absolute order control.

Lastly, I appreciate using conservative numbers, but also measure WebTransport bidi stream response at [~75ms](https://jsfiddle.net/jib1/uadbcev0/) in Chrome.

**Comment by kixelated** [comment]
Unfortunately you can't just arbitrarily drop frames like that. Video is delta-encoded, so a frame can only be decoded if any frames that it depends on have also been decoded. 

My diagram only showed 4 frames per segment, but in reality there's going to be one I-frame followed by at least 59 P-frames (30fps with 2s GoPs) that all depend on the previous frame. For example, if you deprioritize the 5th frame, then the following 54 frames cannot be decoded until it arrives, so they cannot be higher priority.

Technically, the video could be encoded to support this 6-level scheme, however VQ will suffer and I doubt you could use hardware encoding. My point is that the network layer doesn't get to decide the frame dependency graph.

Ideally the network layer should be able to match the encoding dependencies. It becomes possible to support any encoding structure with a sufficiently large number space, as the application can determine the delivery order of each frame.

**Comment by kixelated** [comment]
I absolutely agree that it would a mess to encode arbitrary dependency trees in an integer space. I'm thinking something simpler, like the top x bits indicate the GoP priority while the bottom x bits indicate the frame priority (within a GoP). Anyway, I'm not actually doing frame-level prioritization, so it's kind of a moot point.

I made a [demo](https://github.com/kixelated/warp-demo) for Warp and specifically wanted to point out [how segments are prioritized](https://github.com/kixelated/warp-demo/blob/main/server/internal/warp/session.go#L236). I'm hosting crude public demo (Chrome only) at https://quic.video/demo but it might be at the moment; I'll have to investigate.

**Comment by jan-ivar** [comment]
See also https://github.com/w3c/webtransport/issues/365#issuecomment-1195600435

**Comment by jan-ivar** [comment]
Meeting:
- Some discussion about whether send priorities should also be signaled. Agreed this was more of an IETF issue that can be left orthogonal to this issue about exposing a local JS send priority API.
- Some concern expressed that the discussion has strayed from the OP need for rudimentary prioritization of long-lived streams (e.g. datagram vs data stream vs control-channel) vs. order my (frame-per-stream) superchunks. Would separating into a fresh issue help?

**Comment by jan-ivar** [comment]
With #438 merged the remaining issue here is #451. Closing.
